onnxruntime/contrib_ops/cpu/bert/attention.cc

```diff
@@ -0,0 +1,268 @@
+  }
+
+  const Tensor* weights = context->Input<Tensor>(1);
+  const auto weights_dims = weights->Shape().GetDims();
+  if (weights_dims.size() != 2) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input 1 is expected to have 2 dimensions, got ", weights_dims.size());
+  }
+  if (weights_dims[0] != dims[2]) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input 1 dimension 0 should have same length as dimension 2 of input 0");
+  }
+  if (weights_dims[1] != 3 * weights_dims[0]) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input 1 dimension 1 should be 3 times of dimension 0");
+  }
+
+  const Tensor* bias = context->Input<Tensor>(2);
+  const auto bias_dims = bias->Shape().GetDims();
+  if (bias_dims.size() != 1) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input 2 is expected to have 1 dimension, got ", bias_dims.size());
+  }
+  if (bias_dims[0] != weights_dims[1]) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input 2 dimension 0 should have same length as dimension 1 of input 1");
+  }
+
+  const Tensor* mask_index = context->Input<Tensor>(3);
+  const auto mask_dims = mask_index->Shape().GetDims();
+  if (mask_dims.size() != 1) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Input 3 is expected to have 1 dimension, got ", mask_dims.size());
+  }
+  if (static_cast<int>(mask_dims[0]) != batch_size) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Inputs 3 and 0 shall have same length at dimension 0");
+  }
+
+  return Status::OK();
+}
+
+template <typename T>
+Attention<T>::Attention(const OpKernelInfo& info) : OpKernel(info), AttentionBase(info) {}
+
+template <typename T>
+Status Attention<T>::Compute(OpKernelContext* context) const {
+  ORT_RETURN_IF_ERROR(CheckInputs(context));
+
+  const Tensor* input = context->Input<Tensor>(0);
+  const Tensor* weights = context->Input<Tensor>(1);
+  const Tensor* bias = context->Input<Tensor>(2);
+  const Tensor* mask_index = context->Input<Tensor>(3);
+
+  const auto dims = input->Shape().GetDims();
+  int batch_size = static_cast<int>(dims[0]);
+  int sequence_length = static_cast<int>(dims[1]);
+  int hidden_size = static_cast<int>(dims[2]);
+  int head_size = hidden_size / num_heads_;
+
+  TensorShape output_shape(dims);
+  Tensor* output = context->Output(0, output_shape);
+
+  const size_t element_size = sizeof(T);
+
+  // Use GEMM for fully connection.
+  int m = batch_size * sequence_length;
+  int n = 3 * hidden_size;
+  int k = hidden_size;
+
+  AllocatorPtr allocator;
+  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&allocator));
+
+  concurrency::ThreadPool* tp = context->GetOperatorThreadPool();
```

Comment:
tp can be null.

Reply:
Thanks for the remind. All call paths are checked.